README.md
## parallax-content

Create captivating parallax contents effortlessly with the ParallaxContent plugin. This lightweight Vanilla JavaScript plugin, powered by GSAP animation, supports Scroll and Gyroscope events.

Features:

* Vanilla JavaScript and GSAP powered
* Scroll, Gyroscope events
* Customizable shift and animation duration

### Demo

[Basic demo](https://codesandbox.io/p/devbox/parallax-content-basic-demo-pk4crm)

### Package Managers

```sh
# NPM
npm install parallax_content
```

### Installation

#### Include js

```html
<script src="https://cdnjs.cloudflare.com/ajax/libs/gsap/3.10.4/gsap.min.js"></script>
<script src="parallaxContent.umd.js"></script>
```

#### Set HTML

```html
<div class="parallax-title">Parallax title</div>
```

#### Call the plugin

```html
<script type="text/javascript">
  //Initialize with Vanilla JavaScript
  new ParallaxContent(document.querySelector('.parallax-title'));
  //Initialize with jQuery
  $(document).ready(function () {
    $('.parallax-title').parallaxContent();
  });
</script>
```

#### In result

```html
<html>
  <head>
    <title>My website</title>
  </head>
  <body>
    <div class="parallax-title">Parallax title</div>

    <script src="https://cdnjs.cloudflare.com/ajax/libs/gsap/3.10.4/gsap.min.js"></script>
    <script src="https://code.jquery.com/jquery-3.7.1.js"></script>
    //optional for jQuery initialize
    <script src="node_modules/parallax_content/dist/parallaxContent.umd.js"></script>

    <script type="text/javascript">
      $(document).ready(function () {
        $('.parallax-title').parallaxContent();
      });
    </script>
    <script type="text/javascript">
      //Initialize with Vanilla JavaScript
      new ParallaxContent(document.querySelector('.parallax-title'));
      //Initialize with jQuery
      $(document).ready(function () {
        $('.parallax-title').parallaxContent();
      });
    </script>
  </body>
</html>
```

### Data Attribute Settings

In parallaxContent you can add settings using the data-parallax-content attribute. You still need to call
new ParallaxContent(selector)
to initialize parallaxContent on the element.

Example:

```html
<div data-parallax-content='{"shift": 10, "duration": 4}'></div>
```

### Settings

| Option          | Type  | Default            |
| --------------- | ----- | ------------------ |
| events          | [SCROLL, GYRO] | [SCROLL] |
| shift           | int   | 10                 |
| duration        | int   | 1.5                |
| gyroSensitivity | int   | 30                 |

### Browser support

* Chrome
* Firefox

### Dependencies

* GSAP animation library (Version 3.10.4)

## Contributing

If you'd like to get involved, please consider opening an issue or submitting a pull request. Your input is highly valued, and I'm enthusiastic about collaborating to enhance this tool.

## License

parallax-content is released under the MIT License. See the [LICENSE](LICENSE) file for comprehensive details regarding the terms and conditions of use.


CHANGELOG.md
## [1.0.7] - 2018-7-24

### Added

- Gyro support


gulpfile.js
var 	gulp         = require('gulp'),
        autoprefixer = require('gulp-autoprefixer'),
        minifycss      = require('gulp-uglifycss'),
        uglify       = require('gulp-uglify'),
        rename       = require('gulp-rename'),
        concat       = require('gulp-concat'),
        sass         = require('gulp-sass'),
        plumber      = require('gulp-plumber'),
        notify       = require('gulp-notify'),
        streamqueue  = require('streamqueue'),
        sourcemaps   = require('gulp-sourcemaps'),
        merge = require('merge-stream'),
        babel = require('gulp-babel');



// define the default task and add the watch task to it
gulp.task('default', ['watch']);

gulp.task('styles', function () {

    gulp.task('styles', function () {
        return gulp.src('./demo/sass/style.scss')
            .pipe(plumber({
                errorHandler: notify.onError("Error: <%= error.message %>")
            }))
            .pipe(sourcemaps.init())
            .pipe(sass())
            .pipe(autoprefixer({
                browsers: ['last 5 versions'],
                cascade: false
            }))
            .pipe(sourcemaps.write('/'))
            .pipe(gulp.dest('./demo'))
            .pipe(notify("Styles task complete"));
    });

});


gulp.task('scripts', function() {
    return streamqueue({ objectMode: true },
        gulp.src('./src/parallaxContent.es6')
    )

        .pipe(plumber({
            errorHandler: notify.onError("Error: <%= error.message %>")
        }))

        .pipe(sourcemaps.init())
        .pipe(babel({
            presets: ['es2015']
        }))

        .pipe(gulp.dest('./build'))
        
        .pipe(uglify())
        .pipe(rename({ suffix: '.min' }))
        .pipe(sourcemaps.write('./'))
        .pipe(gulp.dest('./build'))
        .pipe(notify("Custom script task complete"))
});

// configure which files to watch and what tasks to use on file changes
gulp.task('watch', ['styles', 'scripts'], function() {
    gulp.watch('demo/sass/**/*.scss', ['styles']);
    gulp.watch('src/parallaxContent.es6', ['scripts']);
});

demo/webpack.config.js
const autoprefixer = require('autoprefixer');

const NODE_ENV = process.env.NODE_ENV || "development";
const webpack = require('webpack');

module.exports = {
    watch: NODE_ENV == 'development',
    entry: ['./index.es6'],
    output: {
        filename: './build/app.js'
    },
    module: {
        rules: [
            {
                test: /\.es6$/,
                exclude: /(node_modules|bower_components)/,
                use: {
                    loader: 'babel-loader',
                    options: {
                        presets: ['env']
                    }
                }
            },
            {
                test: /\.scss$/,
                use: [
                    {
                        loader: "style-loader"
                    },
                    {
                        loader: "css-loader",
                        options: {
                            sourceMap: NODE_ENV == 'development'
                        }
                    },
                    {
                        loader: 'postcss-loader',
                        options: {
                            plugins: [
                                autoprefixer({
                                    browsers:['last 4 version']
                                })
                            ]
                        }
                    },
                    {
                        loader: "sass-loader",
                        options: {
                        }
                    }
                ]
            }

        ]
    },

    devtool: NODE_ENV == 'development' ? "source-map" : false,

    resolve: {
        alias: {
            jquery: "jquery/src/jquery"
        }
    },

    plugins: [
        new webpack.DefinePlugin({
            NODE_ENV: JSON.stringify(NODE_ENV)
        })
    ]
};

src/index.ts
declare global {
  interface JQuery {
    parallaxContent(arg: Partial<OptionsType>): void;
  }
}

declare global {
  interface Window {
    $: JQuery;
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
declare let gsap: any;

import Scroller from '@lemehovskiy/scroller';

enum EventTypes {
  Scroll = 'SCROLL',
  Gyro = 'GYRO',
}

enum DeviceOrientationTypes {
  Landscape = 'LANDSCAPE',
  Portrait = 'PORTRAIT',
}

type OptionsType = {
  events: Array<EventTypes>;
  shift: number;
  duration: number;
  gyroSensitivity: number;
};

type AnimationParamsType = {
  x?: number;
  y?: number;
};

export default class ParallaxContent {
  element: HTMLElement;
  settings: OptionsType;
  elementSize: [number, number];
  shift: number;
  doubleShift: number;
  deviceOrientation: DeviceOrientationTypes | undefined;
  animationLength: number;
  isVisible: boolean;
  positionX: number;
  positionY: number;

  constructor(element: HTMLElement, options?: Partial<OptionsType>) {
    this.settings = {
      events: [EventTypes.Scroll],
      shift: 30,
      duration: 1.5,
      gyroSensitivity: 30,
      ...options,
    };

    this.element = element;
    const dataOptions = JSON.parse(
      this.element.getAttribute('data-parallax-content')
    );
    this.settings = { ...this.settings, ...dataOptions };
    this.elementSize = [0, 0];
    this.deviceOrientation = undefined;
    this.animationLength = -this.settings.shift + this.settings.shift * 2;
    this.isVisible = false;

    this.init();
  }

  private init() {
    if (typeof gsap === 'undefined') {
      console.warn(
        'Gsap library is required... https://gsap.com/docs/v3/Installation/'
      );
      return;
    }

    const observer = new IntersectionObserver(
      (e) => {
        this.isVisible = e[0].isIntersecting;
      },
      { root: null, rootMargin: '0px' }
    );
    observer.observe(this.element);

    new Set(this.settings.events).forEach((event) => {
      if (event === EventTypes.Scroll) {
        this.subscribeScrollEvent();
      } else if (event === EventTypes.Gyro) {
        this.subscribeGyroEvent();
      }
    });
  }

  private updateOrientation() {
    if (window.innerWidth > window.innerHeight) {
      this.deviceOrientation = DeviceOrientationTypes.Landscape;
    } else {
      this.deviceOrientation = DeviceOrientationTypes.Portrait;
    }
  }

  private subscribeGyroEvent() {
    this.updateOrientation();

    let lastGamma = 0,
      lastBeta = 0,
      rangeGamma = this.settings.gyroSensitivity / 2,
      rangeBeta = this.settings.gyroSensitivity / 2;

    window.addEventListener('resize', () => {
      this.updateOrientation();
    });

    window.addEventListener(
      'deviceorientation',
      (e) => {
        if (!this.isVisible) return;
        const roundedGamma = Math.round(e.gamma || 0),
          roundedBeta = Math.round(e.beta || 0);

        if (roundedBeta > lastBeta && rangeBeta > 0) {
          rangeBeta--;
        } else if (
          roundedBeta < lastBeta &&
          rangeBeta < this.settings.gyroSensitivity
        ) {
          rangeBeta++;
        }

        if (roundedGamma > lastGamma && rangeGamma > 0) {
          rangeGamma--;
        } else if (
          roundedGamma < lastGamma &&
          rangeGamma < this.settings.gyroSensitivity
        ) {
          rangeGamma++;
        }

        lastGamma = roundedGamma;
        lastBeta = roundedBeta;

        const gammaProgress = rangeGamma / this.settings.gyroSensitivity;
        const betaProgress = rangeBeta / this.settings.gyroSensitivity;

        if (this.deviceOrientation === DeviceOrientationTypes.Landscape) {
          this.animate(gammaProgress, betaProgress);
        } else {
          this.animate(betaProgress, gammaProgress);
        }
      },
      true
    );
  }

  private animate(progressY: number, progressX?: number, isSet = false) {
    const y = this.animationLength * progressY;
    const x = this.animationLength * progressX;
    const params: AnimationParamsType = { y };

    if (progressX !== undefined) {
      params.x = x;
    }

    if (isSet) {
      gsap.killTweensOf(this.element);
      gsap.set(this.element, params);
    } else {
      gsap.to(this.element, this.settings.duration, params);
    }
  }

  private subscribeScrollEvent() {
    const scroller = new Scroller(this.element, {
      autoAdjustScrollOffset: true,
    });

    scroller.progressChanged.on((progress) => {
      if (this.isVisible) {
        this.animate(progress);
      } else {
        this.animate(progress, undefined, true);
      }
    });
  }
}

if (window.$ !== undefined) {
  $.fn.parallaxContent = function (
    ...params: [Partial<OptionsType>] | Array<string>
  ) {
    const opt = params[0];
    const args = Array.prototype.slice.call(params, 1);
    const length = this.length;
    let ret = undefined;
    for (let i = 0; i < length; i++) {
      if (typeof opt === 'object' || typeof opt === 'undefined') {
        this[i].parallaxContent = new ParallaxContent(this[i], opt);
      } else {
        // eslint-disable-next-line prefer-spread
        ret = this[i].parallaxContent[opt].apply(this[i].parallaxContent, args);
      }
      if (typeof ret !== 'undefined') return ret;
    }
    return this;
  };
}
